File: src/core/society/institution.ts
// src/core/society/institution.ts
import crypto from 'crypto';
import { ulid } from 'ulid';
import { RuntimeKernel } from '../../kernel/runtime-kernel';
import { DeleteProtection } from '../../data/delete_protection'; // 🔒 Synchronized link back to core protection rules
import { logger } from '../logger';

export type InstitutionScope = 'global' | 'agent' | 'task' | 'domain';
export type EnforcementType = 'hard' | 'soft' | 'advisory';

export interface Institution {
  id: string;
  name: string;
  description: string;
  rules: string[];
  scope: InstitutionScope;
  enforcement: EnforcementType;
  priority: number;
  metadata: Record<string, any>;
  createdAt: number;
  updatedAt: number;
  deletedAt: number | null;
}

/**
 * 🧱 Hardened Constitutional Institution Engine
 * Responsibility: Governs system-wide multi-agent behavioral norms under strict transaction boundaries.
 * Design Spec: Eradicates raw mutable singletons to block multi-instance memory cross-contamination.
 */
export class InstitutionEngine {
  private isOperational = false;
  private readonly moduleName = 'InstitutionEngine';
  private institutions: Map<string, Institution> = new Map();
  private deleteProtection = new DeleteProtection();

  constructor(private kernel: RuntimeKernel) {
    if (!kernel || !kernel.transactionManager || !kernel.commandBus || !kernel.configCenter) {
      throw new Error('CRITICAL_SF_CONSTITUTION: Institution system dependencies missing from bootstrap container.');
    }
  }

  /**
   * 🔌 Component Lifecycle Bootstrapper
   */
  public async boot(): Promise<void> {
    if (this.isOperational) return;

    this.initializeDefaultConstitutionalNorms();

    // Register atomic mutation chains onto central CommandBus
    this.kernel.commandBus.registerHandler('CREATE_SOCIETY_INSTITUTION', async (command: any) => {
      return this.handleCreateInstitutionTransaction(command);
    });

    this.kernel.commandBus.registerHandler('SOFT_DELETE_INSTITUTION', async (command: any) => {
      return this.handleDeleteInstitutionTransaction(command);
    });

    this.isOperational = true;
    logger.info(this.moduleName, '🧱 [OS Phase 3 Institution Core] Hardened norm enforcement engine armed.');
  }

  private initializeDefaultConstitutionalNorms(): void {
    this.registerInternalNormNode({
      name: 'CodeInstitution',
      description: 'Code alterations must traverse rigorous asynchronous review pipelines.',
      rules: ['All code changes require review logs.', 'Reviewers must stay isolated from authors.'],
      scope: 'global', enforcement: 'hard', priority: 100
    });

    this.registerInternalNormNode({
      name: 'SecurityInstitution',
      description: 'High-risk disruptive operations enforce two-man dual validation interlocks.',
      rules: ['Destructive file mutations require secondary confirmations.', 'Anomalous actions trip warning metrics.'],
      scope: 'global', enforcement: 'hard', priority: 150
    });
  }

  private registerInternalNormNode(data: Omit<Institution, 'id' | 'createdAt' | 'updatedAt' | 'deletedAt' | 'metadata'>): void {
    const id = `inst_${ulid()}`;
    const now = Date.now();
    this.institutions.set(id, {
      id, ...data, metadata: {}, createdAt: now, updatedAt: now, deletedAt: null
    });
  }

  /**
   * 🏗️ Command Handler: Two-Phase Version Asserted Norm Record Appender
   */
  private async handleCreateInstitutionTransaction(command: any): Promise<Institution> {
    const { traceId, name, description, rules, scope, enforcement, priority, metadata } = command.payload;
    const initialVersion = this.kernel.version;

    const tx = await this.kernel.transactionManager.begin(
      command.id || crypto.randomUUID(),
      this.moduleName,
      { traceId, name, enforcement, readVersionStamp: initialVersion }
    );

    try {
      if (this.kernel.version !== initialVersion) {
        throw new Error(`ERR_SF_INSTITUTION_RACE: Version mismatch during cold norm blueprint compilation.`);
      }

      const id = `inst_${ulid()}`;
      const now = Date.now();
      const institutionBlock: Institution = {
        id, name, description, rules: rules || [], scope, enforcement,
        priority: Number(priority || 100), metadata: metadata || {},
        createdAt: now, updatedAt: now, deletedAt: null
      };

      this.institutions.set(id, institutionBlock);

      tx.payload = {
        ...tx.payload,
        institution_id: id,
        norm_name_seal: name,
        scope_rank: scope,
        priority_coefficient: institutionBlock.priority,
        finalized_at: now
      };

      await this.kernel.transactionManager.commit(tx.id);
      this.pushMetrics('society.institution.norms_created', 1);
      return institutionBlock;

    } catch (panic: any) {
      await this.kernel.transactionManager.rollback(tx.commandId, panic);
      this.pushMetrics('society.institution.failures_count', 1);
      throw panic;
    }
  }

  /**
   * 🏗️ Command Handler: Hardened Soft-Delete Transactional Controller
   */
  private async handleDeleteInstitutionTransaction(command: any): Promise<boolean> {
    const { traceId, institutionId } = command.payload;
    const targetNorm = this.institutions.get(institutionId);
    if (!targetNorm || targetNorm.deletedAt) return false;

    // Rigid check mapping over core protection frameworks prior to initiating state locks
    const check = this.deleteProtection.canDelete('institution', institutionId);
    if (!check.allowed) {
      logger.error(this.moduleName, `🚨 Unauthorized deletion attempt intercepted for norm entry: ${institutionId}. Reason: ${check.reason}`);
      return false;
    }

    const initialVersion = this.kernel.version;
    const tx = await this.kernel.transactionManager.begin(
      command.id || crypto.randomUUID(),
      this.moduleName,
      { traceId, institutionId, readVersionStamp: initialVersion }
    );

    try {
      if (this.kernel.version !== initialVersion) {
        throw new Error(`ERR_SF_INSTITUTION_DELETE_RACE: Version drifted prior to soft-deleting target norm asset.`);
      }

      const now = Date.now();
      targetNorm.deletedAt = now;
      targetNorm.updatedAt = now;

      tx.payload = {
        ...tx.payload,
        institution_id: institutionId,
        deletion_status_seal: 'soft_deleted',
        finalized_at: now
      };

      await this.kernel.transactionManager.commit(tx.id);
      this.pushMetrics('society.institution.norms_deleted', 1);
      return true;

    } catch (panic: any) {
      await this.kernel.transactionManager.rollback(tx.commandId, panic);
      throw panic;
    }
  }

  public getNormProxy(id: string): Institution | undefined {
    const inst = this.institutions.get(id);
    return inst && !inst.deletedAt ? { ...inst } : undefined;
  }

  public getEffectiveNormRules(): Institution[] {
    return Array.from(this.institutions.values())
      .filter(i => !i.deletedAt)
      .sort((x, y) => y.priority - x.priority);
  }

  private pushMetrics(metricName: string, value: number) {
    if (this.kernel?.metricsCollector?.counter) {
      this.kernel.metricsCollector.counter(metricName, value, { domain: 'society', layer: 'institution' });
    }
  }

  public evictEngineCache(): void {
    this.institutions.clear();
    this.isOperational = false;
  }
}

File: python/soloforge_ai_society/models/institution.py
# -*- coding: utf-8 -*-
"""
SoloForge AI Society - Institution（制度系统）

制度是行为规范的结构化集合，定义 AI 社会的规则体系。
"""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import List, Optional
import uuid


class InstitutionScope(Enum):
    """制度生效范围"""
    GLOBAL = "global"      # 全局生效
    AGENT = "agent"        # 仅对特定 Agent
    TASK = "task"          # 仅对特定任务
    DOMAIN = "domain"      # 仅对特定领域


class EnforcementType(Enum):
    """执行类型"""
    HARD = "hard"          # 强制执行
    SOFT = "soft"          # 软性引导
    ADVISORY = "advisory"  # 建议提示


@dataclass
class Institution:
    """
    制度系统

    属性：
        id: 唯一标识符
        name: 制度名称（如 "CodeInstitution"）
        rules: 规则列表
        scope: 生效范围
        enforcement: 执行类型
        priority: 优先级（冲突时高优先级覆盖）
        created_at: 创建时间
        updated_at: 更新时间
    """

    name: str
    rules: List[str]
    scope: InstitutionScope = InstitutionScope.GLOBAL
    enforcement: EnforcementType = EnforcementType.HARD
    priority: int = 50
    id: str = field(default_factory=lambda: f"inst_{uuid.uuid4().hex[:12]}")
    created_at: datetime = field(default_factory=datetime.now)
    updated_at: datetime = field(default_factory=datetime.now)

    # 可选范围限定
    agent_id: Optional[str] = None
    task_type: Optional[str] = None
    domain: Optional[str] = None

    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "id": self.id,
            "name": self.name,
            "rules": self.rules,
            "scope": self.scope.value,
            "enforcement": self.enforcement.value,
            "priority": self.priority,
            "agent_id": self.agent_id,
            "task_type": self.task_type,
            "domain": self.domain,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Institution":
        """从字典创建"""
        return cls(
            id=data["id"],
            name=data["name"],
            rules=data["rules"],
            scope=InstitutionScope(data["scope"]),
            enforcement=EnforcementType(data["enforcement"]),
            priority=data.get("priority", 50),
            agent_id=data.get("agent_id"),
            task_type=data.get("task_type"),
            domain=data.get("domain"),
            created_at=datetime.fromisoformat(data["created_at"]),
            updated_at=datetime.fromisoformat(data["updated_at"]),
        )


# 预置制度
PRESET_INSTITUTIONS = [
    Institution(
        id="inst_core_code_review",
        name="CodeInstitution",
        rules=[
            "代码修改必须经过 Review",
            "Review 必须有书面反馈",
            "高风险变更需要多人确认",
        ],
        scope=InstitutionScope.GLOBAL,
        enforcement=EnforcementType.HARD,
        priority=90,
    ),
    Institution(
        id="inst_core_research",
        name="ResearchInstitution",
        rules=[
            "研究结论必须有证据链",
            "引用来源必须可追溯",
            "不确定时必须标注置信度",
        ],
        scope=InstitutionScope.DOMAIN,
        enforcement=EnforcementType.SOFT,
        priority=70,
        domain="research",
    ),
    Institution(
        id="inst_core_security",
        name="SecurityInstitution",
        rules=[
            "高风险操作必须双人确认",
            "敏感操作必须记录审计日志",
            "未经确认禁止删除数据",
        ],
        scope=InstitutionScope.GLOBAL,
        enforcement=EnforcementType.HARD,
        priority=95,
    ),
]
